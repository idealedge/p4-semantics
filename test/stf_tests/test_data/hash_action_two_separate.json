{
  "program" : "hash_action_two_separate.p4",
  "__meta__" : {
    "version" : [2, 7],
    "compiler" : "https://github.com/p4lang/p4c"
  },
  "header_types" : [
    {
      "name" : "scalars_0",
      "id" : 0,
      "fields" : [
        ["tmp", 32, false],
        ["tmp_0", 32, false]
      ]
    },
    {
      "name" : "data_t",
      "id" : 1,
      "fields" : [
        ["f1", 32, false],
        ["f2", 32, false],
        ["h1", 16, false],
        ["h2", 16, false],
        ["h3", 16, false],
        ["h4", 16, false]
      ]
    },
    {
      "name" : "counter_metadata_t",
      "id" : 2,
      "fields" : [
        ["counter_index_first", 16, false],
        ["counter_index_second", 16, false]
      ]
    },
    {
      "name" : "standard_metadata",
      "id" : 3,
      "fields" : [
        ["ingress_port", 9, false],
        ["egress_spec", 9, false],
        ["egress_port", 9, false],
        ["clone_spec", 32, false],
        ["instance_type", 32, false],
        ["drop", 1, false],
        ["recirculate_port", 16, false],
        ["packet_length", 32, false],
        ["enq_timestamp", 32, false],
        ["enq_qdepth", 19, false],
        ["deq_timedelta", 32, false],
        ["deq_qdepth", 19, false],
        ["ingress_global_timestamp", 48, false],
        ["lf_field_list", 32, false],
        ["mcast_grp", 16, false],
        ["resubmit_flag", 1, false],
        ["egress_rid", 16, false],
        ["checksum_error", 1, false],
        ["_padding", 4, false]
      ]
    }
  ],
  "headers" : [
    {
      "name" : "scalars",
      "id" : 0,
      "header_type" : "scalars_0",
      "metadata" : true,
      "pi_omit" : true
    },
    {
      "name" : "standard_metadata",
      "id" : 1,
      "header_type" : "standard_metadata",
      "metadata" : true,
      "pi_omit" : true
    },
    {
      "name" : "data",
      "id" : 2,
      "header_type" : "data_t",
      "metadata" : false,
      "pi_omit" : true
    },
    {
      "name" : "counter_metadata",
      "id" : 3,
      "header_type" : "counter_metadata_t",
      "metadata" : true,
      "pi_omit" : true
    }
  ],
  "header_stacks" : [],
  "header_union_types" : [],
  "header_unions" : [],
  "header_union_stacks" : [],
  "field_lists" : [],
  "errors" : [
    ["NoError", 1],
    ["PacketTooShort", 2],
    ["NoMatch", 3],
    ["StackOutOfBounds", 4],
    ["HeaderTooShort", 5],
    ["ParserTimeout", 6]
  ],
  "enums" : [],
  "parsers" : [
    {
      "name" : "parser",
      "id" : 0,
      "init_state" : "start",
      "parse_states" : [
        {
          "name" : "start",
          "id" : 0,
          "parser_ops" : [
            {
              "parameters" : [
                {
                  "type" : "regular",
                  "value" : "data"
                }
              ],
              "op" : "extract"
            }
          ],
          "transitions" : [
            {
              "value" : "default",
              "mask" : null,
              "next_state" : null
            }
          ],
          "transition_key" : []
        }
      ]
    }
  ],
  "deparsers" : [
    {
      "name" : "deparser",
      "id" : 0,
      "order" : ["data"]
    }
  ],
  "meter_arrays" : [],
  "counter_arrays" : [
    {
      "name" : "count1",
      "id" : 0,
      "size" : 16384,
      "is_direct" : false
    },
    {
      "name" : "count2",
      "id" : 1,
      "size" : 16384,
      "is_direct" : false
    }
  ],
  "register_arrays" : [],
  "calculations" : [],
  "learn_lists" : [],
  "actions" : [
    {
      "name" : "NoAction",
      "id" : 0,
      "runtime_data" : [],
      "primitives" : []
    },
    {
      "name" : "set_index",
      "id" : 1,
      "runtime_data" : [
        {
          "name" : "index1",
          "bitwidth" : 16
        },
        {
          "name" : "index2",
          "bitwidth" : 16
        },
        {
          "name" : "port",
          "bitwidth" : 9
        }
      ],
      "primitives" : [
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["counter_metadata", "counter_index_first"]
            },
            {
              "type" : "runtime_data",
              "value" : 0
            }
          ],
          "source_info" : {
            "filename" : "hash_action_two_separate.p4",
            "line" : 29,
            "column" : 17,
            "source_fragment" : "index1, index2, port) { ..."
          }
        },
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["counter_metadata", "counter_index_second"]
            },
            {
              "type" : "runtime_data",
              "value" : 1
            }
          ],
          "source_info" : {
            "filename" : "hash_action_two_separate.p4",
            "line" : 29,
            "column" : 25,
            "source_fragment" : "index2, port) { ..."
          }
        },
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["standard_metadata", "egress_spec"]
            },
            {
              "type" : "runtime_data",
              "value" : 2
            }
          ],
          "source_info" : {
            "filename" : "hash_action_two_separate.p4",
            "line" : 29,
            "column" : 33,
            "source_fragment" : "port) { ..."
          }
        }
      ]
    },
    {
      "name" : "count_entries",
      "id" : 2,
      "runtime_data" : [],
      "primitives" : [
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["scalars", "tmp"]
            },
            {
              "type" : "expression",
              "value" : {
                "type" : "expression",
                "value" : {
                  "op" : "&",
                  "left" : {
                    "type" : "field",
                    "value" : ["counter_metadata", "counter_index_first"]
                  },
                  "right" : {
                    "type" : "hexstr",
                    "value" : "0xffffffff"
                  }
                }
              }
            }
          ]
        },
        {
          "op" : "count",
          "parameters" : [
            {
              "type" : "counter_array",
              "value" : "count1"
            },
            {
              "type" : "field",
              "value" : ["scalars", "tmp"]
            }
          ],
          "source_info" : {
            "filename" : "hash_action_two_separate.p4",
            "line" : 61,
            "column" : 4,
            "source_fragment" : "count(count1, counter_metadata.counter_index_first)"
          }
        }
      ]
    },
    {
      "name" : "count_entries2",
      "id" : 3,
      "runtime_data" : [],
      "primitives" : [
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["scalars", "tmp_0"]
            },
            {
              "type" : "expression",
              "value" : {
                "type" : "expression",
                "value" : {
                  "op" : "&",
                  "left" : {
                    "type" : "field",
                    "value" : ["counter_metadata", "counter_index_second"]
                  },
                  "right" : {
                    "type" : "hexstr",
                    "value" : "0xffffffff"
                  }
                }
              }
            }
          ]
        },
        {
          "op" : "count",
          "parameters" : [
            {
              "type" : "counter_array",
              "value" : "count2"
            },
            {
              "type" : "field",
              "value" : ["scalars", "tmp_0"]
            }
          ],
          "source_info" : {
            "filename" : "hash_action_two_separate.p4",
            "line" : 72,
            "column" : 4,
            "source_fragment" : "count(count2, counter_metadata.counter_index_second)"
          }
        }
      ]
    }
  ],
  "pipelines" : [
    {
      "name" : "ingress",
      "id" : 0,
      "init_table" : "index_setter",
      "tables" : [
        {
          "name" : "index_setter",
          "id" : 0,
          "source_info" : {
            "filename" : "hash_action_two_separate.p4",
            "line" : 35,
            "column" : 0,
            "source_fragment" : "table index_setter { ..."
          },
          "key" : [
            {
              "match_type" : "exact",
              "target" : ["data", "f1"],
              "mask" : null
            },
            {
              "match_type" : "exact",
              "target" : ["data", "f2"],
              "mask" : null
            }
          ],
          "match_type" : "exact",
          "type" : "simple",
          "max_size" : 2048,
          "with_counters" : false,
          "support_timeout" : false,
          "direct_meters" : null,
          "action_ids" : [1, 0],
          "actions" : ["set_index", "NoAction"],
          "base_default_next" : "stats",
          "next_tables" : {
            "set_index" : "stats",
            "NoAction" : "stats"
          },
          "default_entry" : {
            "action_id" : 0,
            "action_const" : false,
            "action_data" : [],
            "action_entry_const" : false
          }
        },
        {
          "name" : "stats",
          "id" : 1,
          "source_info" : {
            "filename" : "hash_action_two_separate.p4",
            "line" : 64,
            "column" : 0,
            "source_fragment" : "table stats { ..."
          },
          "key" : [],
          "match_type" : "exact",
          "type" : "simple",
          "max_size" : 1024,
          "with_counters" : false,
          "support_timeout" : false,
          "direct_meters" : null,
          "action_ids" : [2],
          "actions" : ["count_entries"],
          "base_default_next" : "stats2",
          "next_tables" : {
            "count_entries" : "stats2"
          },
          "default_entry" : {
            "action_id" : 2,
            "action_const" : false,
            "action_data" : [],
            "action_entry_const" : false
          }
        },
        {
          "name" : "stats2",
          "id" : 2,
          "source_info" : {
            "filename" : "hash_action_two_separate.p4",
            "line" : 75,
            "column" : 0,
            "source_fragment" : "table stats2 { ..."
          },
          "key" : [],
          "match_type" : "exact",
          "type" : "simple",
          "max_size" : 1024,
          "with_counters" : false,
          "support_timeout" : false,
          "direct_meters" : null,
          "action_ids" : [3],
          "actions" : ["count_entries2"],
          "base_default_next" : null,
          "next_tables" : {
            "count_entries2" : null
          },
          "default_entry" : {
            "action_id" : 3,
            "action_const" : false,
            "action_data" : [],
            "action_entry_const" : false
          }
        }
      ],
      "action_profiles" : [],
      "conditionals" : []
    },
    {
      "name" : "egress",
      "id" : 1,
      "init_table" : null,
      "tables" : [],
      "action_profiles" : [],
      "conditionals" : []
    }
  ],
  "checksums" : [],
  "force_arith" : [],
  "extern_instances" : [],
  "field_aliases" : [
    [
      "queueing_metadata.enq_timestamp",
      ["standard_metadata", "enq_timestamp"]
    ],
    [
      "queueing_metadata.enq_qdepth",
      ["standard_metadata", "enq_qdepth"]
    ],
    [
      "queueing_metadata.deq_timedelta",
      ["standard_metadata", "deq_timedelta"]
    ],
    [
      "queueing_metadata.deq_qdepth",
      ["standard_metadata", "deq_qdepth"]
    ],
    [
      "intrinsic_metadata.ingress_global_timestamp",
      ["standard_metadata", "ingress_global_timestamp"]
    ],
    [
      "intrinsic_metadata.lf_field_list",
      ["standard_metadata", "lf_field_list"]
    ],
    [
      "intrinsic_metadata.mcast_grp",
      ["standard_metadata", "mcast_grp"]
    ],
    [
      "intrinsic_metadata.resubmit_flag",
      ["standard_metadata", "resubmit_flag"]
    ],
    [
      "intrinsic_metadata.egress_rid",
      ["standard_metadata", "egress_rid"]
    ]
  ]
}