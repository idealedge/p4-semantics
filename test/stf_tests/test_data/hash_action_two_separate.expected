00010001000100010001000100010001001000100010001000100010001000100000000100000001000000100000001000000011000000110000010000000100 1
00100010001000100010001000100010001100110011001100110011001100110000001000000010000000110000001100000100000001000000010100000101 2
00110011001100110011001100110011010001000100010001000100010001000000001100000011000001000000010000000101000001010000011000000110 2
01000100010001000100010001000100010101010101010101010101010101010000010000000100000001010000010100000110000001100000011100000111 3